src/main/java/org/openlmis/requisition/service/RequisitionService.java

```diff
@@ -807,6 +808,72 @@ private boolean isRequisitionNewest(Requisition requisition) {
     return null == recentRequisition || requisition.getId().equals(recentRequisition.getId());
   }
 
+
+  /**
+   * Returns statistics data regarding requisitions statuses for a specified facility:
+   * number of requisitions to be created for current periods and number of requisitions
+   * with each status available in the system.
+   *
+   * @param facility FacilityDto object.
+   * @return RequisitionStatsData object.
+   */
+  public RequisitionStatsData getStatusesStatsData(FacilityDto facility) {
+    UUID facilityId = facility.getId();
+    List<RequisitionStatus> postSubmittedStatuses = new ArrayList<>();
+    Map<String, Long> statusesStats = new HashMap<>();
+    for (RequisitionStatus status : RequisitionStatus.values()) {
+      statusesStats.put(
+          status.name(),
+          requisitionRepository.countRequisitions(null, facilityId,
+              null, null, status)
+      );
+      if (status.isPostSubmitted()) {
+        postSubmittedStatuses.add(status);
+      }
+    }
+    RequisitionStatsData requisitionStatsData = new RequisitionStatsData();
+    requisitionStatsData.setStatusesStats(statusesStats);
+    requisitionStatsData.setFacilityId(facilityId);
+
+    // The maximum number of combinations of actively supported programs and periods for a given
+    // facility will be the number of requisitions that can be created for that facility.
+    Set<ProcessingPeriodDto> currentFacilityPeriods = new HashSet<>();
+    List<SupportedProgramDto> activelySupportedPrograms = facility.getSupportedPrograms()
+        .stream()
+        .filter(SupportedProgramDto::isSupportActive)
+        .filter(SupportedProgramDto::isProgramActive)
+        .collect(toList());
+    activelySupportedPrograms.forEach(program -> {
+      if (program.isSupportActive()) {
+        List<ProcessingPeriodDto> foundPeriods =
+            new ArrayList<>(periodService.searchByProgramAndFacilityAndDateRange(
+                program.getId(), facilityId,
+                LocalDate.now(), LocalDate.now()));
+        currentFacilityPeriods.addAll(foundPeriods);
+      }
+    });
+
+    long maxRequisitionForFacility = currentFacilityPeriods.size();
+    if (maxRequisitionForFacility == 0) {
+      requisitionStatsData.setRequisitionsToBeCreated(0L);
+    } else {
+      List<UUID> activelySupportedProgramsIds = activelySupportedPrograms.stream()
+          .map(SupportedProgramDto::getId)
+          .collect(toList());
+      List<UUID> currentFacilityPeriodsIds = currentFacilityPeriods.stream()
+          .map(ProcessingPeriodDto::getId)
+          .collect(toList());
+      Long createdRequisitions = requisitionRepository.countRequisitions(
+          facilityId, activelySupportedProgramsIds, currentFacilityPeriodsIds, null,
+          postSubmittedStatuses
+      );
+      long requisitionsToBeCreated = maxRequisitionForFacility - createdRequisitions;
+      requisitionStatsData.setRequisitionsToBeCreated(requisitionsToBeCreated);
+    }
```

Comment:
This function is too long, it will be great to extract three individual ones out of it.
I would also consider adding three profilers, one for each method usage.

Reply:
the method doesn't seem that long compared to others in this class, however I extracted some lines of code into separate methods to make the code look cleaner and added profilers where it made sense.